docs/software_architecture.md

```diff
@@ -0,0 +1,52 @@
+# Software architecture
+
+## HTML rewriting
+
+HTML rewriting is purely static (i.e. before resources are written to the ZIM). HTML code is parsed with the [HTML parser from Python standard library](https://docs.python.org/3/library/html.parser.html).
+
+Some code is statically injected with a [Jinja2](https://pypi.org/project/Jinja2/) template used to populate some JS context variables needed by wombat operations (original scheme, original url, ...).
+
+## CSS rewriting
+
+CSS rewriting is purely static (i.e. before resources are written to the ZIM). CSS code is parsed with the [tinycss2 Python library](https://pypi.org/project/tinycss2/).
+
+## JS rewriting
+
+### Static
+
+Static JS rewriting is simply a matter of pure textual manipulation with regular expressions. No parsing is done at all.
+
+### Dynamic
+
+Dynamic JS rewriting is done with [wombat JS library](https://github.com/webrecorder/wombat). The same fuzzy rules that are used for
+static rewritting are injected into wombat configuration. Code to rewrite URLs is an adapted version of the code used to compute ZIM paths.
+
+## cdxj_indexer and warcio
+
+[cdxj_indexer Python library](https://pypi.org/project/cdxj-indexer/) is a thin wrapper over [warcio Python library](https://pypi.org/project/warcio/). It is used to efficiently iterate over WARC resources and their contents. Some operations are directly done with warcio when the wrapper is useless.
+
+**TODO** confirm last statements
+
+## chardet
+
+[chardet Python library](https://pypi.org/project/chardet/) is used to detect character encoding of files when it is absent (only HTML file typically specify its encoding) or incoherent.
+
+## zimscraperlib
+
+[zimscraperlib Python library](https://pypi.org/project/zimscraperlib) is used for ZIM operations.
+
+## requests
+
+[requests Python library](https://pypi.org/project/requests/) is used to retrieve the custom CSS file when a URL is passed.
+
+## Babel
+
+**TODO** seems unused, could be removed?
+
+## brotlipy
+
+**TODO** seems unused, could be removed?
+
+## beautifulsoup4
+
+**TODO** seems unused, could be removed?
```

Comment:
Yes, it was used in warc2zim 1 but now we extract information (title) when rewriting html.